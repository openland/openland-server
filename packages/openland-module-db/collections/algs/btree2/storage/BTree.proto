syntax = "proto3";

message TreeNode {
    required int32 id = 0;
    optional int32 parent = 1;
    required TreeNodeType type = 2;
    repeated int32 values = 3;
    repeated NodeChildren children = 4;
}

message NodeChildren {
    required int32 id = 0;
    required int32 max = 1;
    required int32 min = 2;
    required int32 count = 3;
}

message TreeHead {
    required int32 counter = 1;
    optional int32 root = 2;
}

enum TreeNodeType {
    LEAF = 0;
    INNER = 1;
}