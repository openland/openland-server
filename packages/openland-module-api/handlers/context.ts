import * as express from 'express';
import { Modules } from 'openland-modules/Modules';
import { createTracer } from 'openland-log/createTracer';
import { createLogger } from 'openland-log/createLogger';
import { createEmptyContext } from 'openland-utils/Context';
import { AuthContext } from 'openland-module-auth/AuthContext';
import { TracingContext } from 'openland-log/TracingContext';
import { CacheContext } from 'openland-module-api/CacheContext';
import { AppContext } from 'openland-modules/AppContext';

let tracer = createTracer('express');
const logger = createLogger('http');

async function context(src: express.Request): Promise<AppContext> {

    let res = createEmptyContext();
    let uid: number | undefined;
    let tid: string | undefined;
    let oid: number | undefined;

    // User
    if (src.user !== null && src.user !== undefined) {
        if (typeof src.user.sub === 'string') {
            uid = await Modules.Users.findUserByAuthId(createEmptyContext(), src.user.sub);
        } else if (typeof src.user.uid === 'number' && typeof src.user.tid === 'number') {
            uid = src.user.uid;
            tid = src.user.tid;
        }
    }
    // Organization
    if (uid) {
        let accounts = await Modules.Orgs.findUserOrganizations(createEmptyContext(), uid);

        // Default behaviour: pick the default one
        if (accounts.length >= 1) {
            oid = accounts[0];

            let profile = await Modules.Users.profileById(createEmptyContext(), uid);
            oid = (profile && profile.primaryOrganization) || oid;
        }
    }

    // Auth Context
    res = AuthContext.set(res, { tid, uid, oid });

    // Tracing Context
    res = TracingContext.set(res, { span: tracer.startSpan('http') });
    res = CacheContext.set(res, new Map());
    return new AppContext(res);
}

export async function callContextMiddleware(isTest: boolean, req: express.Request, res: express.Response, next: express.NextFunction) {
    let ctx: AppContext;
    try {
        ctx = await context(req);
    } catch (e) {
        res!!.status(404).send('Unable to find domain');
        return;
    }
    if (!isTest) {
        if (AuthContext.get(ctx).uid) {
            logger.log('GraphQL [#' + AuthContext.get(ctx).uid + ']: ' + JSON.stringify(req.body));
        } else {
            logger.log('GraphQL [#ANON]: ' + JSON.stringify(req.body));
        }
    }
    res.locals.ctx = ctx;

    const originalEnd = res.end;
    res.end = function (...args: any[]) {
        res.end = originalEnd;
        const returned = res.end.call(this, ...args);
        let span = TracingContext.get(ctx).span;
        if (span) {
            span.finish();
        }
        return returned;
    };

    next();
}