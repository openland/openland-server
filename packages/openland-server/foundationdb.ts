import { Shutdown } from './../openland-utils/Shutdown';
import * as fs from 'fs';
import { Database, Layer } from '@openland/foundationdb';
import migrations from './migrations';
import { RandomLayer } from '@openland/foundationdb-random';
import { MigrationsLayer } from '@openland/foundationdb-migrations';
import { LockLayer } from '@openland/foundationdb-locks';
import { SingletonWorkerLayer } from '@openland/foundationdb-singleton';
import { BusLayer, NoOpBus } from '@openland/foundationdb-bus';
import { RedisBusProvider } from '@openland/foundationdb-bus-redis';

function createLayers(test: boolean) {
    let layers: Layer[] = [
        new RandomLayer(),
        new MigrationsLayer(migrations),
        new LockLayer(),
        new SingletonWorkerLayer(),
        new BusLayer(
            !process.env.REDIS_HOST
                ? new NoOpBus()
                : new RedisBusProvider(
                    process.env.REDIS_PORT ? parseInt(process.env.REDIS_PORT as string, 10) : 6379,
                    process.env.REDIS_HOST
                ))
    ];
    return layers;
}

export async function openDatabase() {
    let db: Database;
    if (process.env.FOUNDATION_DB) {
        fs.writeFileSync('foundation.clusterfile', process.env.FOUNDATION_DB);
        db = await Database.open({ clusterFile: 'foundation.clusterfile', layers: createLayers(false) });
    } else {
        db = await Database.open({ layers: createLayers(false) });
    }
    Shutdown.registerWork({
        name: 'database',
        shutdown: async (ctx) => {
            await db.close(ctx);
        }
    });
    return db;
}

export async function openTestDatabase() {
    let db: Database;
    if (process.env.FOUNDATION_DB) {
        fs.writeFileSync('foundation.clusterfile', process.env.FOUNDATION_DB);
        db = await Database.openTest({ clusterFile: 'foundation.clusterfile', layers: createLayers(true) });
    } else {
        db = await Database.openTest({ layers: createLayers(true) });
    }
    Shutdown.registerWork({
        name: 'database',
        shutdown: async (ctx) => {
            await db.close(ctx);
        }
    });
    return db;
}