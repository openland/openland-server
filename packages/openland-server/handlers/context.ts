import * as express from 'express';
import { CallContext } from '../api/utils/CallContext';
import { Repos } from '../repositories';
import { Modules } from 'openland-modules/Modules';

async function context(src: express.Request): Promise<CallContext> {
    let res = new CallContext();
    //
    // Loading UID
    //
    if (src.user !== null && src.user !== undefined) {
        if (typeof src.user.sub === 'string') {
            res.uid = await Repos.Users.fetchUserByAuthId(src.user.sub);
        } else if (typeof src.user.uid === 'number' && typeof src.user.tid === 'number') {
            res.uid = src.user.uid;
            res.tid = src.user.tid;
        }
    }

    res.ip = src.ip;

    //
    // Loading Organization
    //
    if (res.uid) {
        let accounts = await Repos.Users.fetchUserAccounts(res.uid);

        // Default behaviour: pick the default one
        if (accounts.length >= 1) {
            res.oid = accounts[0];

            let profile = await Modules.Users.profileById(res.uid);
            res.oid = (profile && profile.primaryOrganization) || res.oid;
        }

        // res.superRope = await Repos.Permissions.superRole(res.uid);
    }

    return res;
}

export async function callContextMiddleware(isTest: boolean, req: express.Request, res: express.Response, next: express.NextFunction) {
    let ctx: CallContext;
    try {
        ctx = await context(req);
    } catch (e) {
        res!!.status(404).send('Unable to find domain');
        return;
    }
    if (!isTest) {
        if (ctx.uid) {
            console.log('GraphQL [#' + ctx.uid + ']: ' + JSON.stringify(req.body));
        } else {
            console.log('GraphQL [#ANON]: ' + JSON.stringify(req.body));
        }
    }
    res.locals.ctx = ctx;
    next();
}