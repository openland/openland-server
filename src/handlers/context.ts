import * as express from 'express';
import { CallContext } from '../api/utils/CallContext';
import { Repos } from '../repositories';
import { IDs } from '../api/utils/IDs';
import { fetchKeyFromRequest } from '../utils/fetchKeyFromRequest';

async function context(src: express.Request): Promise<CallContext> {
    let res = new CallContext();
    //
    // Loading UID
    //
    if (src.user !== null && src.user !== undefined) {
        if (typeof src.user.sub === 'string') {
            res.uid = await Repos.Users.fetchUserByAuthId(src.user.sub);
        } else if (typeof src.user.id === 'number') {
            res.uid = src.user.id;
        }
    }

    //
    // Loading Organization
    //
    if (res.uid) {
        let accounts = await Repos.Users.fetchUserAccounts(res.uid);

        // Default behaviour: pick the default one
        if (accounts.length === 1) {
            res.oid = accounts[0];
        }

        // If there are organization cookie, try to use it instead
        let orgId = fetchKeyFromRequest(src, 'x-openland-org');
        if (orgId) {
            if (Array.isArray(orgId)) {
                orgId = orgId[0];
            }
            try {
                let porgId = IDs.Organization.parse(orgId as string);
                if (accounts.indexOf(porgId) >= 0) {
                    res.oid = porgId;
                }
            } catch (e) {
                console.warn(e);
            }
        }
    }

    return res;
}

export async function callContextMiddleware(isTest: boolean, req: express.Request, res: express.Response, next: express.NextFunction) {
    let ctx: CallContext;
    try {
        ctx = await context(req);
    } catch (e) {
        res!!.status(404).send('Unable to find domain');
        return;
    }
    if (!isTest) {
        if (ctx.uid) {
            console.log('GraphQL [#' + ctx.uid + ']: ' + JSON.stringify(req.body));
        } else {
            console.log('GraphQL [#ANON]: ' + JSON.stringify(req.body));
        }
    }
    res.locals.ctx = ctx;
    next();
}