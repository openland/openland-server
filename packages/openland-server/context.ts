import { CacheContext } from 'openland-module-api/CacheContext';
import { AuthContext } from 'openland-module-auth/AuthContext';
import { ContextNamespaceType, registerExtension } from '@openland/context';
import { RequestContext } from 'openland-module-api/RequestContext';

declare module '@openland/context' {
    interface Context {
        readonly auth: ContextNamespaceType<typeof AuthContext>;
        readonly req: ContextNamespaceType<typeof RequestContext>;
        readonly cache: Map<string, any>;
    }
}

registerExtension('auth', (ctx) => {
    return AuthContext.get(ctx);
});
registerExtension('req', (ctx) => {
    return RequestContext.get(ctx);
});
registerExtension('cache', (ctx) => {
    return CacheContext.get(ctx)!;
});