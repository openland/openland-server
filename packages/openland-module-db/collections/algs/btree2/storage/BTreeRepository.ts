import { Context } from '@openland/context';
import { encoders, getTransaction, Subspace, TransactionCache, TupleItem } from '@openland/foundationdb';
import { TreeHead, TreeNode, TreeNodeType } from './BTree';

type TreeCache = {
    head: {
        root: number,
        nodeCounter: number,
    },
    cache: { [key: number]: TreeNode | null },
    writes: { [key: number]: true }
    writeHead: boolean,
};

const bucketWriteCache = new TransactionCache<TreeCache>('btree-repo');

const SUBSPACE_HEAD = 0;
const SUBSPACE_NODES = 1;

export class BTreeRepository {
    private readonly subspace: Subspace<TupleItem[], Buffer>;

    constructor(subspace: Subspace) {
        this.subspace = subspace.withKeyEncoding(encoders.tuple);
    }

    //
    // Nodes
    //

    async readNode(ctx: Context, collection: Buffer, node: number) {
        let cache = await this.getCache(ctx, collection);
        let cached = cache.cache[node];
        if (cached !== undefined) {
            if (cached === null) {
                throw Error('Unable to find node ' + node);
            }
            return cached;
        }
        let rawNode = await this.subspace.get(ctx, [collection, SUBSPACE_NODES, node]);
        if (!rawNode) {
            throw Error('Unable to find node ' + node);
        }
        let decoded = TreeNode.decode(rawNode);
        cache.cache[node] = decoded;
        return decoded;
    }

    async clearNode(ctx: Context, collection: Buffer, id: number) {
        let cache = await this.getCache(ctx, collection);
        let hadWrites = cache.writeHead || Object.keys(cache.writes).length > 0;
        cache.cache[id] = null;
        cache.writes[id] = true;
        if (!hadWrites) {
            this.registerFlush(ctx, collection, cache);
        }
    }

    async writeNode(ctx: Context, collection: Buffer, node: TreeNode) {
        if (node.type === TreeNodeType.INNER && node.children.length < 2) {
            throw Error('Invalid inner node');
        }
        let cache = await this.getCache(ctx, collection);
        let hadWrites = cache.writeHead || Object.keys(cache.writes).length > 0;
        cache.cache[node.id] = node;
        cache.writes[node.id] = true;
        if (!hadWrites) {
            this.registerFlush(ctx, collection, cache);
        }
    }

    async allocateNodeId(ctx: Context, collection: Buffer) {
        let cache = await this.getCache(ctx, collection);
        let id = ++cache.head.nodeCounter;
        this.flushHead(ctx, collection, cache);
        return id;
    }

    async getRoot(ctx: Context, collection: Buffer) {
        let cache = await this.getCache(ctx, collection);
        if (cache.head.root > 0) {
            return cache.head.root;
        } else {
            return null;
        }
    }

    async setRoot(ctx: Context, collection: Buffer, node: number | null) {
        let cache = await this.getCache(ctx, collection);
        if (node === null) {
            cache.head.root = 0;
        } else {
            cache.head.root = node;
        }
        this.flushHead(ctx, collection, cache);
    }

    //
    // Cache operations
    //

    private flushHead(ctx: Context, collection: Buffer, cache: TreeCache) {
        if (!cache.writeHead) {
            cache.writeHead = true;

            // If not already flushed
            if (Object.keys(cache.writes).length === 0) {
                this.registerFlush(ctx, collection, cache);
            }
        }
    }

    private registerFlush(ctx: Context, collection: Buffer, cache: TreeCache) {
        getTransaction(ctx).beforeCommit((commit) => {

            // Flush head
            if (cache.writeHead) {
                this.subspace.set(commit, [collection, SUBSPACE_HEAD], Buffer.from(TreeHead.encode({ root: cache.head.root, counter: cache.head.nodeCounter }).finish()));
            }

            // Flush nodes
            for (let key in cache.writes) {
                let id = parseInt(key, 10);
                if (cache.cache[id]) {
                    this.subspace.set(commit, [collection, SUBSPACE_NODES, id], Buffer.from(TreeNode.encode(cache.cache[id]!).finish()));
                } else {
                    this.subspace.clear(commit, [collection, SUBSPACE_NODES, id]);
                }
            }
        });
    }

    private async getCache(ctx: Context, collection: Buffer) {
        let key = 'cache-' + collection.toString('hex');
        let existing = bucketWriteCache.get(ctx, key);
        if (!existing) {
            let headRaw = await this.subspace.get(ctx, [collection, SUBSPACE_HEAD]);
            if (!headRaw) {
                existing = { head: { nodeCounter: 1, root: 0 }, writeHead: false, cache: {}, writes: {} };
                bucketWriteCache.set(ctx, key, existing);
            } else {
                let decoded = TreeHead.decode(headRaw);
                existing = { head: { nodeCounter: decoded.counter, root: decoded.root }, writeHead: false, cache: {}, writes: {} };
                bucketWriteCache.set(ctx, key, existing);
            }
        }
        return existing;
    }
}