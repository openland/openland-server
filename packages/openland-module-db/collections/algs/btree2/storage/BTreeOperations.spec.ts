import { createNamedContext } from '@openland/context';
import { inTx, Database } from '@openland/foundationdb';
import { TreeNodeType } from './BTree';
import { BTreeRepository } from './BTreeRepository';
import { BTreeOperations } from './BTreeOperations';

const COLLECTION_0 = Buffer.from([0]);
const COLLECTION_1 = Buffer.from([1]);
const COLLECTION_2 = Buffer.from([2]);
const COLLECTION_3 = Buffer.from([3]);
const COLLECTION_4 = Buffer.from([4]);
const COLLECTION_5 = Buffer.from([5]);
const COLLECTION_6 = Buffer.from([6]);

describe('BTreeOperations', () => {
    let db: Database;
    let root = createNamedContext('test');
    let ops: BTreeOperations;

    beforeAll(async () => {
        db = await Database.openTest({ name: 'bplustree-directory-ops', layers: [] });
        ops = new BTreeOperations(new BTreeRepository(db.allKeys));
    });

    it('should create and delete root node', async () => {
        let dump = await inTx(root, async (ctx) => {
            await ops.setRootNode(ctx, COLLECTION_0, { id: 1, type: TreeNodeType.LEAF, values: [1, 2, 3, 4] });
            return await ops.dumpAll(ctx, COLLECTION_0);
        });
        expect(dump).toMatchSnapshot();

        dump = await inTx(root, async (ctx) => {
            await ops.removeValuesFromLeaf(ctx, COLLECTION_0, 1, [1, 3]);
            return await ops.dumpAll(ctx, COLLECTION_0);
        });
        expect(dump).toMatchSnapshot();
    });

    it('should split root node', async () => {
        await inTx(root, async (ctx) => {
            await ops.setRootNode(ctx, COLLECTION_1, { id: 1, type: TreeNodeType.LEAF, values: [1, 2, 3, 4] });
        });

        let dump = await inTx(root, async (ctx) => {
            await ops.splitNode(ctx, COLLECTION_1, 1);
            return await ops.dumpAll(ctx, COLLECTION_1);
        });
        expect(dump).toMatchSnapshot();

        dump = await inTx(root, async (ctx) => {
            await ops.splitNode(ctx, COLLECTION_1, 1);
            return await ops.dumpAll(ctx, COLLECTION_1);
        });
        expect(dump).toMatchSnapshot();

        dump = await inTx(root, async (ctx) => {
            await ops.splitNode(ctx, COLLECTION_1, 3);
            return await ops.dumpAll(ctx, COLLECTION_1);
        });
        expect(dump).toMatchSnapshot();

        dump = await inTx(root, async (ctx) => {
            await ops.splitNode(ctx, COLLECTION_1, 2);
            return await ops.dumpAll(ctx, COLLECTION_1);
        });
        expect(dump).toMatchSnapshot();
    });

    it('should find node', async () => {
        await inTx(root, async (ctx) => {
            await ops.setRootNode(ctx, COLLECTION_2, { id: 1, type: TreeNodeType.LEAF, values: [1, 2, 3, 4, 5, 6, 7, 8] });
            await ops.splitNode(ctx, COLLECTION_2, 1);
            await ops.splitNode(ctx, COLLECTION_2, 1);
            await ops.splitNode(ctx, COLLECTION_2, 3);
            await ops.splitNode(ctx, COLLECTION_2, 2);
        });

        await inTx(root, async (ctx) => {
            expect((await ops.search(ctx, COLLECTION_2, 1)).id).toBe(1);
            expect((await ops.search(ctx, COLLECTION_2, 2)).id).toBe(1);
            expect((await ops.search(ctx, COLLECTION_2, 3)).id).toBe(4);
            expect((await ops.search(ctx, COLLECTION_2, 4)).id).toBe(4);
            expect((await ops.search(ctx, COLLECTION_2, 5)).id).toBe(3);
            expect((await ops.search(ctx, COLLECTION_2, 6)).id).toBe(3);
            expect((await ops.search(ctx, COLLECTION_2, 7)).id).toBe(5);
            expect((await ops.search(ctx, COLLECTION_2, 8)).id).toBe(5);

            expect((await ops.findNodeSiblingLeft(ctx, COLLECTION_2, 1))).toBeNull();
            expect((await ops.findNodeSiblingRight(ctx, COLLECTION_2, 1))).toBe(4);
            expect((await ops.findNodeSiblingLeft(ctx, COLLECTION_2, 4))).toBe(1);
            expect((await ops.findNodeSiblingRight(ctx, COLLECTION_2, 4))).toBe(3);
            expect((await ops.findNodeSiblingLeft(ctx, COLLECTION_2, 3))).toBe(4);
            expect((await ops.findNodeSiblingRight(ctx, COLLECTION_2, 3))).toBe(5);
            expect((await ops.findNodeSiblingLeft(ctx, COLLECTION_2, 5))).toBe(3);
            expect((await ops.findNodeSiblingRight(ctx, COLLECTION_2, 5))).toBeNull();
        });
    });

    it('should insert to leaf', async () => {
        await inTx(root, async (ctx) => {
            await ops.setRootNode(ctx, COLLECTION_3, { id: 1, type: TreeNodeType.LEAF, values: [1, 2, 3, 4, 5, 6, 7, 8] });
            await ops.splitNode(ctx, COLLECTION_3, 1);
            await ops.splitNode(ctx, COLLECTION_3, 1);
            await ops.splitNode(ctx, COLLECTION_3, 3);
        });

        let dump = await inTx(root, async (ctx) => {
            await ops.addValuesToLeaf(ctx, COLLECTION_3, 1, [0, -1]);
            await ops.addValuesToLeaf(ctx, COLLECTION_3, 5, [10, 9]);
            return ops.dumpAll(ctx, COLLECTION_3);
        });
        expect(dump).toMatchSnapshot();
    });

    it('should delete node', async () => {
        await inTx(root, async (ctx) => {
            await ops.setRootNode(ctx, COLLECTION_4, { id: 1, type: TreeNodeType.LEAF, values: [1, 2, 3, 4, 5, 6, 7, 8] });
            await ops.splitNode(ctx, COLLECTION_4, 1);
            await ops.splitNode(ctx, COLLECTION_4, 1);
            await ops.splitNode(ctx, COLLECTION_4, 3);
        });

        let dump = await inTx(root, async (ctx) => {
            await ops.deleteNode(ctx, COLLECTION_4, 1);
            return ops.dumpAll(ctx, COLLECTION_4);
        });
        expect(dump).toMatchSnapshot();
    });

    it('should move value', async () => {
        await inTx(root, async (ctx) => {
            await ops.setRootNode(ctx, COLLECTION_5, { id: 1, type: TreeNodeType.LEAF, values: [1, 2, 3, 4] });
            await ops.splitNode(ctx, COLLECTION_5, 1);
        });

        let dump = await inTx(root, async (ctx) => {
            await ops.moveValue(ctx, COLLECTION_5, 1, 3, 2);
            return ops.dumpAll(ctx, COLLECTION_5);
        });
        expect(dump).toMatchSnapshot();

        dump = await inTx(root, async (ctx) => {
            await ops.moveValue(ctx, COLLECTION_5, 3, 1, 2);
            await ops.moveValue(ctx, COLLECTION_5, 3, 1, 3);
            return ops.dumpAll(ctx, COLLECTION_5);
        });
        expect(dump).toMatchSnapshot();
    });

    it('should move node', async () => {
        let dump = await inTx(root, async (ctx) => {
            await ops.setRootNode(ctx, COLLECTION_6, { id: 1, type: TreeNodeType.LEAF, values: [1, 2, 3, 4, 5, 6, 7, 8] });
            await ops.splitNode(ctx, COLLECTION_6, 1);
            await ops.splitNode(ctx, COLLECTION_6, 1);
            await ops.splitNode(ctx, COLLECTION_6, 3);
            await ops.splitNode(ctx, COLLECTION_6, 2);
            return ops.dumpAll(ctx, COLLECTION_6);
        });
        expect(dump).toMatchSnapshot();

        dump = await inTx(root, async (ctx) => {
            await ops.moveNode(ctx, COLLECTION_6, 4, 7);
            return ops.dumpAll(ctx, COLLECTION_6);
        });
        expect(dump).toMatchSnapshot();
    });
});