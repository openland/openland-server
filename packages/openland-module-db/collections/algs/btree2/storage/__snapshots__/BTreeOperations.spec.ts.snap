// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BTreeOperations should create and delete root node 1`] = `
Object {
  "id": 1,
  "parent": null,
  "type": "leaf",
  "values": Array [
    1,
    2,
    3,
    4,
  ],
}
`;

exports[`BTreeOperations should create and delete root node 2`] = `
Object {
  "id": 1,
  "parent": null,
  "type": "leaf",
  "values": Array [
    2,
    4,
  ],
}
`;

exports[`BTreeOperations should delete node 1`] = `
Object {
  "children": Array [
    Object {
      "count": 2,
      "max": 4,
      "min": 3,
      "node": Object {
        "id": 4,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          3,
          4,
        ],
      },
    },
    Object {
      "count": 2,
      "max": 6,
      "min": 5,
      "node": Object {
        "id": 3,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          5,
          6,
        ],
      },
    },
    Object {
      "count": 2,
      "max": 8,
      "min": 7,
      "node": Object {
        "id": 5,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          7,
          8,
        ],
      },
    },
  ],
  "id": 2,
  "parent": null,
  "type": "internal",
}
`;

exports[`BTreeOperations should insert to leaf 1`] = `
Object {
  "children": Array [
    Object {
      "count": 4,
      "max": 2,
      "min": -1,
      "node": Object {
        "id": 1,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          -1,
          0,
          1,
          2,
        ],
      },
    },
    Object {
      "count": 2,
      "max": 4,
      "min": 3,
      "node": Object {
        "id": 4,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          3,
          4,
        ],
      },
    },
    Object {
      "count": 2,
      "max": 6,
      "min": 5,
      "node": Object {
        "id": 3,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          5,
          6,
        ],
      },
    },
    Object {
      "count": 4,
      "max": 10,
      "min": 7,
      "node": Object {
        "id": 5,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          7,
          8,
          9,
          10,
        ],
      },
    },
  ],
  "id": 2,
  "parent": null,
  "type": "internal",
}
`;

exports[`BTreeOperations should move node 1`] = `
Object {
  "children": Array [
    Object {
      "count": 4,
      "max": 4,
      "min": 1,
      "node": Object {
        "children": Array [
          Object {
            "count": 2,
            "max": 2,
            "min": 1,
            "node": Object {
              "id": 1,
              "parent": 2,
              "type": "leaf",
              "values": Array [
                1,
                2,
              ],
            },
          },
          Object {
            "count": 2,
            "max": 4,
            "min": 3,
            "node": Object {
              "id": 4,
              "parent": 2,
              "type": "leaf",
              "values": Array [
                3,
                4,
              ],
            },
          },
        ],
        "id": 2,
        "parent": 6,
        "type": "internal",
      },
    },
    Object {
      "count": 4,
      "max": 8,
      "min": 5,
      "node": Object {
        "children": Array [
          Object {
            "count": 2,
            "max": 6,
            "min": 5,
            "node": Object {
              "id": 3,
              "parent": 7,
              "type": "leaf",
              "values": Array [
                5,
                6,
              ],
            },
          },
          Object {
            "count": 2,
            "max": 8,
            "min": 7,
            "node": Object {
              "id": 5,
              "parent": 7,
              "type": "leaf",
              "values": Array [
                7,
                8,
              ],
            },
          },
        ],
        "id": 7,
        "parent": 6,
        "type": "internal",
      },
    },
  ],
  "id": 6,
  "parent": null,
  "type": "internal",
}
`;

exports[`BTreeOperations should move node 2`] = `
Object {
  "children": Array [
    Object {
      "count": 2,
      "max": 2,
      "min": 1,
      "node": Object {
        "children": Array [
          Object {
            "count": 2,
            "max": 2,
            "min": 1,
            "node": Object {
              "id": 1,
              "parent": 2,
              "type": "leaf",
              "values": Array [
                1,
                2,
              ],
            },
          },
        ],
        "id": 2,
        "parent": 6,
        "type": "internal",
      },
    },
    Object {
      "count": 6,
      "max": 8,
      "min": 3,
      "node": Object {
        "children": Array [
          Object {
            "count": 2,
            "max": 4,
            "min": 3,
            "node": Object {
              "id": 4,
              "parent": 7,
              "type": "leaf",
              "values": Array [
                3,
                4,
              ],
            },
          },
          Object {
            "count": 2,
            "max": 6,
            "min": 5,
            "node": Object {
              "id": 3,
              "parent": 7,
              "type": "leaf",
              "values": Array [
                5,
                6,
              ],
            },
          },
          Object {
            "count": 2,
            "max": 8,
            "min": 7,
            "node": Object {
              "id": 5,
              "parent": 7,
              "type": "leaf",
              "values": Array [
                7,
                8,
              ],
            },
          },
        ],
        "id": 7,
        "parent": 6,
        "type": "internal",
      },
    },
  ],
  "id": 6,
  "parent": null,
  "type": "internal",
}
`;

exports[`BTreeOperations should move value 1`] = `
Object {
  "children": Array [
    Object {
      "count": 1,
      "max": 1,
      "min": 1,
      "node": Object {
        "id": 1,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          1,
        ],
      },
    },
    Object {
      "count": 3,
      "max": 4,
      "min": 2,
      "node": Object {
        "id": 3,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          2,
          3,
          4,
        ],
      },
    },
  ],
  "id": 2,
  "parent": null,
  "type": "internal",
}
`;

exports[`BTreeOperations should move value 2`] = `
Object {
  "children": Array [
    Object {
      "count": 3,
      "max": 3,
      "min": 1,
      "node": Object {
        "id": 1,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          1,
          2,
          3,
        ],
      },
    },
    Object {
      "count": 1,
      "max": 4,
      "min": 4,
      "node": Object {
        "id": 3,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          4,
        ],
      },
    },
  ],
  "id": 2,
  "parent": null,
  "type": "internal",
}
`;

exports[`BTreeOperations should split root node 1`] = `
Object {
  "children": Array [
    Object {
      "count": 2,
      "max": 2,
      "min": 1,
      "node": Object {
        "id": 1,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          1,
          2,
        ],
      },
    },
    Object {
      "count": 2,
      "max": 4,
      "min": 3,
      "node": Object {
        "id": 3,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          3,
          4,
        ],
      },
    },
  ],
  "id": 2,
  "parent": null,
  "type": "internal",
}
`;

exports[`BTreeOperations should split root node 2`] = `
Object {
  "children": Array [
    Object {
      "count": 1,
      "max": 1,
      "min": 1,
      "node": Object {
        "id": 1,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          1,
        ],
      },
    },
    Object {
      "count": 1,
      "max": 2,
      "min": 2,
      "node": Object {
        "id": 4,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          2,
        ],
      },
    },
    Object {
      "count": 2,
      "max": 4,
      "min": 3,
      "node": Object {
        "id": 3,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          3,
          4,
        ],
      },
    },
  ],
  "id": 2,
  "parent": null,
  "type": "internal",
}
`;

exports[`BTreeOperations should split root node 3`] = `
Object {
  "children": Array [
    Object {
      "count": 1,
      "max": 1,
      "min": 1,
      "node": Object {
        "id": 1,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          1,
        ],
      },
    },
    Object {
      "count": 1,
      "max": 2,
      "min": 2,
      "node": Object {
        "id": 4,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          2,
        ],
      },
    },
    Object {
      "count": 1,
      "max": 3,
      "min": 3,
      "node": Object {
        "id": 3,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          3,
        ],
      },
    },
    Object {
      "count": 1,
      "max": 4,
      "min": 4,
      "node": Object {
        "id": 5,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          4,
        ],
      },
    },
  ],
  "id": 2,
  "parent": null,
  "type": "internal",
}
`;

exports[`BTreeOperations should split root node 4`] = `
Object {
  "children": Array [
    Object {
      "count": 2,
      "max": 2,
      "min": 1,
      "node": Object {
        "children": Array [
          Object {
            "count": 1,
            "max": 1,
            "min": 1,
            "node": Object {
              "id": 1,
              "parent": 2,
              "type": "leaf",
              "values": Array [
                1,
              ],
            },
          },
          Object {
            "count": 1,
            "max": 2,
            "min": 2,
            "node": Object {
              "id": 4,
              "parent": 2,
              "type": "leaf",
              "values": Array [
                2,
              ],
            },
          },
        ],
        "id": 2,
        "parent": 6,
        "type": "internal",
      },
    },
    Object {
      "count": 2,
      "max": 4,
      "min": 3,
      "node": Object {
        "children": Array [
          Object {
            "count": 1,
            "max": 3,
            "min": 3,
            "node": Object {
              "id": 3,
              "parent": 7,
              "type": "leaf",
              "values": Array [
                3,
              ],
            },
          },
          Object {
            "count": 1,
            "max": 4,
            "min": 4,
            "node": Object {
              "id": 5,
              "parent": 7,
              "type": "leaf",
              "values": Array [
                4,
              ],
            },
          },
        ],
        "id": 7,
        "parent": 6,
        "type": "internal",
      },
    },
  ],
  "id": 6,
  "parent": null,
  "type": "internal",
}
`;
